### PR TITLE
[#811] Revoke host docker socket from model pods

### LINKED ISSUE #811: Revoke host docker socket from model pods
We are creating model images with container commit call with using docker socket from host node.
Sharing docker socket with model pod is not secure so we should consider alternatives for this process.

### PR DESCRIPTION
**Pull request by mcdoker18**
* Added new parameter for the legionctl command: `--build-type`.
  Currently, it supports the following values:
    * `docker-socket` - builds a model image using a docker socket. It
      is the default value.
    * `docker-remote` - does not build image yourself. Send an HTTP
      request to the `MODEL_DOCKER_BUILDER_URL` server that determines
      the model container, build and pushes it.
    * `kaniko` or `buildah` will be supported in the near future.
* Added the second container to slave model Jenkins pod that will build
a model docker image.

This closes #811

### CODE REVIEW

File: k8s/jenkins/legion-jenkins-plugin/src/main/resources/legion.groovy [hunk 1]
```diff
@@ -112,12 +114,28 @@ spec:
...
+                            command: '/usr/local/bin/uwsgi',
+                            serviceAccount: 'jenkins',
+                            args: "--http 127.0.0.1:8080 --module legion.containers.wsgi --processes 2",
+                            envVars: envVars,
+                            volumes: [
+                                    hostPathVolume(hostPath: '/var/run/docker.sock', mountPath: '/var/run/docker.sock')
+                            ])
             ],
             volumes: [
                     hostPathVolume(hostPath: '/var/run/docker.sock', mountPath: '/var/run/docker.sock'),
```
Comment (kirillmakhonin) [inline k8s/jenkins/legion-jenkins-plugin/src/main/resources/legion.groovy]: Why?

File: k8s/jenkins/legion-jenkins-plugin/src/main/resources/legion.groovy [hunk 2]
```diff
@@ -112,12 +114,28 @@ spec:
...
-                            image: image,
+                            image: image_model,
                             resourceLimitMemory: ram,
                             resourceLimitCpu: cpu,
                             ttyEnabled: true,
                             command: 'cat',
-                            envVars: envVars),
+                            envVars: envVars,
+                            volumes: [
+                                    persistentVolumeClaim(claimName: env.DAGS_VOLUME_PVC, mountPath: env.AIRFLOW_DAGS_DIRECTORY)
```
Comment (kirillmakhonin) [inline k8s/jenkins/legion-jenkins-plugin/src/main/resources/legion.groovy]: Please remove

File: legion/legion/cli.py [hunk 3]
```diff
@@ -88,34 +88,26 @@ def build_model(args):
...
-    legion.utils.send_header_to_stderr(legion.containers.headers.IMAGE_ID_LOCAL, image.id)
+    params = legion.containers.docker.ModelDockerBuildParameters(model_id, workspace_path, image_labels,
+                                                                 new_image_tag, args.push_to_registry)
 
-    if image.tags:
-        legion.utils.send_header_to_stderr(legion.containers.headers.IMAGE_TAG_LOCAL, image.tags[0])
-
-    if args.push_to_registry:
-        legion.containers.docker.push_image_to_registry(client, image, args.push_to_registry)
+    if args.build_type == 'docker-socket':
```
Comment (kirillmakhonin) [inline legion/legion/cli.py]: Please use constants or Enums

File: legion/legion/cli.py [hunk 4]
```diff
@@ -580,6 +572,12 @@ def build_parser():  # pylint: disable=R0915
                                     type=str, help='docker image tag')
     build_model_parser.add_argument('--push-to-registry',
                                     type=str, help='docker registry address')
+    build_model_parser.add_argument('--build-type', default='docker-socket',
```
Comment (kirillmakhonin) [inline legion/legion/cli.py]: Please use choices

File: legion/legion/config.py [hunk 5]
```diff
@@ -363,6 +363,11 @@ def __new__(cls, name, default=None, cast_func=str, description=None, configurab
                                       'Name of node [Jenkins]. It is used in model saving (during training)',
                                       False)
 
+MODEL_DOCKER_BUILDER_URL = ConfigVariableDeclaration('MODEL_DOCKER_BUILDER_URL', 'http://127.0.0.1:8080', str,
+                                                     'Model docker builder url', False)
```
Comment (kirillmakhonin) [inline legion/legion/config.py]: Please extend description of new config variables

File: legion/legion/containers/server.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+@blueprint.route(DOCKER_BUILD_URL, methods=['POST'])
+@legion.http.provide_json_response
+def build():
+    """
+    Build API endpoint
+
+    :return: image name
+    """
+    params = ModelDockerBuildParameters(**request.get_json(force=True))
+    model_container_id = get_container_id_of_current_pod('model')
```
Comment (kirillmakhonin) [inline legion/legion/containers/server.py]: Please use constants

File: legion/legion/containers/server.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+from legion.containers.docker import ModelDockerBuildParameters, build_model_docker_image
+from legion.k8s.utils import get_container_id_of_current_pod
+from legion.logging import set_log_level, redirect_to_stdout
+
+LOGGER = logging.getLogger(__name__)
+blueprint = Blueprint('docker_server', __name__)
+TEMPLATES_FOLDER = os.path.abspath(os.path.join(
+    os.path.dirname(__file__), os.pardir, 'templates', 'edi')
+)
+AUTH_COOKIE_NAME = '_oauth2_proxy'
```
Comment (kirillmakhonin) [inline legion/legion/containers/server.py]: Please remove

File: legion/legion/containers/server.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+import legion.http
+import legion.model
+from legion.containers.docker import ModelDockerBuildParameters, build_model_docker_image
+from legion.k8s.utils import get_container_id_of_current_pod
+from legion.logging import set_log_level, redirect_to_stdout
+
+LOGGER = logging.getLogger(__name__)
+blueprint = Blueprint('docker_server', __name__)
+TEMPLATES_FOLDER = os.path.abspath(os.path.join(
```
Comment (kirillmakhonin) [inline legion/legion/containers/server.py]: Redundant

File: legion/legion/containers/wsgi.py [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+"""
+Entry point for WSGI server
+"""
+
+from legion.containers.server import init_application
```
Comment (kirillmakhonin) [inline legion/legion/containers/wsgi.py]: Can we remove this file?

File: legion/legion/external/docker.py [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+
+def request_to_build_image(params: ModelDockerBuildParameters) -> str:
+    """
+    Send request to build a model container
+
+    :param params: Model docker build parameters
+    :return model docker image
+    """
+    resp = requests.post(config.MODEL_DOCKER_BUILDER_URL.strip('/') + DOCKER_BUILD_URL,  # pylint: disable=E1101
```
Comment (kirillmakhonin) [inline legion/legion/external/docker.py]: Please add exception handling

File: legion/legion/external/docker.py [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+                         json=params._asdict())
+    resp.raise_for_status()
+
+    return resp.json()["image_name"]
```
Comment (kirillmakhonin) [inline legion/legion/external/docker.py]: Please use NamedTuple

**Review by kirillmakhonin — CHANGES_REQUESTED** [review]

**Comment by mcdoker18** [comment]
bot, test all

**Comment by mcdoker18** [comment]
bot, test all

**Comment by mcdoker18** [comment]
bot, test all

**Comment by mcdoker18** [comment]
bot, test all

**Review by kirillmakhonin — APPROVED** [review]

**Review by botsman — APPROVED** [review]

**Review by alexander-semenets — APPROVED** [review]

**Pull request merged**